pkg/monitoring/vmstat/hyperv/hyperv_wmi.go

```diff
@@ -0,0 +1,228 @@
+// +build windows
+
+package hyperv
+
+import (
+	"fmt"
+	"net"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/StackExchange/wmi"
+
+	log "github.com/sirupsen/logrus"
+)
+
+var guestNetworkRegexp *regexp.Regexp
+
+func init() {
+	guestNetworkRegexp = regexp.MustCompile(`^Microsoft:GuestNetwork\\(\w{8}-\w{4}-\w{4}-\w{4}-\w{12})\\(\w{8}-\w{4}-\w{4}-\w{4}-\w{12})$`)
+}
+
+type vmStat struct {
+	InstanceName           string
+	CPUWaitTimePerDispatch float64
+	State                  string
+	CPUUsagePercent        float32
+	AssignedMemory         string
+	Uptime                 string
+	IPAddresses            []string
+	Heartbeat              string
+	GUID                   string
+}
+
+type EnabledState uint16
+type HealthState uint16
+type Heartbeat uint16
+
+type msvm_ComputerSystem struct {
+	ElementName           string
+	InstallDate           string
+	HealthState           HealthState
+	EnabledState          EnabledState
+	OtherEnabledState     string
+	RequestedState        EnabledState
+	TimeOfLastStateChange string
+	Name                  string
+	ProcessID             uint32
+}
+
+type msvm_SummaryInformation struct {
+	Name                 string
+	ElementName          string
+	GuestOperatingSystem *string
+	Version              string
+	EnabledState         EnabledState
+	HealthState          HealthState
+	CreationTime         time.Time
+	MemoryUsage          *uint64
+	Uptime               uint64
+	ProcessorLoad        *uint16
+	Heartbeat            *Heartbeat
+	NumberOfProcessors   uint16
+}
+
+type msvm_GuestNetworkAdapterConfiguration struct {
+	InstanceID  string
+	IPAddresses []string
+}
+
+type hypervStat struct {
+	TotalCPUWaitTimePerDispatch float64
+	List                        map[string]vmStat
+}
+
+var hypervPath = "\\Hyper-V Hypervisor Virtual processor(*)\\CPU Wait Time Per Dispatch"
+
+func (im *impl) GetMeasurements() (map[string]interface{}, error) {
+	meas := make(map[string]interface{})
+
+	stat := &hypervStat{
+		List: make(map[string]vmStat),
+	}
+
+	var countersErr error
+	res, countersErr := im.watcher.GetFormattedQueryData(hypervPath)
+	if countersErr == nil {
+		// Iterate over the reported counter values
+		for _, c := range res {
+			// Some filtering...
```

Comment:
If a comment is added, make it more useful :)

Reply:
done